handler.js

```diff
@@ -277,13 +277,57 @@ export function ingest({ Records: records }: Object, context: Object) {
           return parse(raw)
             .then(mapToAirings)
             .then(function(airings) {
-
               console.log('Parsed airings list', airings.length);
 
+              let airingData = {};
+
+              airings.forEach(function (a: Airing) {
+                if (airingData[a.channel] === undefined) {
+                  airingData[a.channel] = {
+                    channel: a.channel,
+                    startTime: 0,
+                    endTime: 0,
+                    airings: []
+                  };
+                }
+                airingData[a.channel].airings.push(a.id);
+                if (airingData[a.channel].endTime === 0 || a.date < airingData[a.channel].startTime) {
+                  airingData[a.channel].startTime = a.date;
+                }
+                if (airingData[a.channel].endTime === 0 || a.end_date > airingData[a.channel].endTime) {
+                  airingData[a.channel].endTime = a.date;
+                }
+              });
+
+              let removeAiringPromises = [];
+
+              for (let a in airingData) {
+                if (airingData.hasOwnProperty(a)) {
+                  a = airingData[a];
+                  let searchPromise = actions.searchAirings({
+                    channel: a.channel,
+                    start: a.startTime,
+                    end: a.endTime
+                  });
+
+                  searchPromise.then(function (channelAirings) {
+                    let diff = channelAirings.filter((airing) => {
+                      return !a.airings.includes(airing.data.id);
+                    });
+
+                    if (diff.length) {
+                      removeAiringPromises.push(actions.removeAirings(diff.map(a => a.data)));
+                    }
+                  });
+                }
+              }
+
               let p1 = actions.insertAirings(airings);
               let p2 = actions.insertShows(airings.map(a => a.show));
 
-              return p1.then(() => p2)
+              return Promise.all(removeAiringPromises)
```

Comment:
This is some of the tricky-ness of promises. The current form doesn't guarantee that the queries run in the order expected, and they may all end up running at the same time. Instead we want something like:

(note: given the above there would only be a single remove promise)

```
removeAiringPromises.then(() => {
  let p1 = actions.insertAirings(airings);
  let p2 = actions.insertShows(airings.map(a => a.show));
  return Promise.all([p1, p2]);
});
```

This can be read as run `removeAiringPromises` first, and once it completes then run `insertAirings` and `insertShows` in parallel. When both are done, resolve the chain.